(TRAINTECH-8595) Enhance Puppet AWS reaper for classroom/SE demo environments

greglarkin opened the pull request:
The ticket https://tickets.puppetlabs.com/browse/TRAINTECH-8595 describes the need for enhancing the AWS reaper to tear down additional resources provisioned by the classroom/SE demo infrastructure tooling.

This is very much "work in progress" code. An additional Lambda function entry point has been added for dev/test purposes now and will need to be integrated properly further down the road. To test this code, do the following:

```
git clone git@github.com:greglarkin/aws_resource_reaper.git
cd aws_resource_reaper
git checkout enhance-aws-reaper
export aws_region=us-west-2 # OPTIONAL: set to desired region
export aws_secret_access_key=CHANGEME
export aws_access_key=CHANGEME
docker run -e AWS_REGION=$aws_region -e AWS_ACCESS_KEY_ID=$aws_access_key -e AWS_SECRET_ACCESS_KEY=$aws_secret_access_key --rm -v "$PWD":/var/task lambci/lambda:python2.7 reaper.terminate_expired_resources
```

At the moment, I'm just starting the code to identify all of the resources (load balancers so far, target groups next, then Route 53 records and VPCs) that need to be deleted. I'm using similar patterns to what was already in place for terminating EC2 instances.

Commit:
Extend the AWS reaper to handle other resource types

Commit:
Starting work on deleting load balancers and target groups

greglarkin commented:
@ShawonC If you have any comments on my current direction with the reaper enhancements, please let me know. It's my first time using Boto3 and my Python skills are pretty beginner, but hopefully it's a good start.

One concern is that some resources depend on others and may need to have a wait loop inserted before proceeding to try deleting a dependent resource.

Oddly, I'm also already running into an issue where some load balancer termination dates look correct to me but cannot be parsed. Other ones can, so maybe there's an issue in the code somewhere that I'm not seeing:

```
Unable to parse the termination_date 2019-05-20 14:55:03 -0700 for load balancer arn:aws:elasticloadbalancing:us-west-2:051431149377:loadbalancer/app/Demo-pierre-qa3-ALB-MASTER/e8eb432c431b94a5
Unable to parse the termination_date 2019-05-20 14:55:03 -0700 for load balancer arn:aws:elasticloadbalancing:us-west-2:051431149377:loadbalancer/app/Demo-pierre-qa3-ALB-WELCOME/8bdbc861ec29076c
...
...
Load balancer arn:aws:elasticloadbalancing:us-west-2:051431149377:loadbalancer/app/GSWP-1964-ALB-WELCOME/27abf9773f5f8323 will be deleted 83899 seconds from now, roughly
Load balancer arn:aws:elasticloadbalancing:us-west-2:051431149377:loadbalancer/app/GSWP-1964-ALB-CD4PE/496e5491946e77fc will be deleted 83899 seconds from now, roughly
```

Commit:
Fixed red herring exception - use named parameters for describing TGs

greglarkin commented:
Nevermind about the parse error I described above. That was a red herring, and I got that code working today.

acidprime commented:
@njbrubaker Based on cost explored this is costing a few thousand a month right now and is likely to start being more of an issue now that the SE demo environment is being used more heavily. Let us know if you want to do a meeting to talk about it more.

Commit:
Added support for classic load balancer and target group deletion

greglarkin commented:
Ran the script for the first time in live mode and called the load balancer entry point function directly. It now enumerates classic and v2 load balancers as well as target groups. TGs will need to be pulled out and enumerated separately, I think. Ran into "resource in use" errors trying to delete them, so need to look into that and possibly deregister them?

Cleaned up LBs in Oregon, N. Virginia, London and Frankfurt so far. Still plenty of code cleanup to do but at least the cost should be reduced while that is being done. Will clean up the other regions as well. /cc @acidprime

Commit:
Added a waiter for terminating EC2 instances and added code to
remove target groups. Unified the Lambda entry point function to
delete all expired resources.

greglarkin commented:
With the latest commit, the new entrypoint function is named `terminate_expired_resources`

greglarkin commented:
Also worth noting - nothing is done with LBs and target groups that have no `termination_date` tag. EC2 instances are stopped when they are detected as improperly tagged. There are a number of other resources that have no tags, but they might be a mix of legacy stuff that we need and other things that are not needed so some manual inspection will be needed.

Commit:
Add code to delete expired subnets

greglarkin commented:
@ShawonC This PR is ready to review, and it terminates load balancers, the other resource that was costing us money beyond EC2 instances. I've started work on cleaning up the other resources that we provision, including:

- subnets (included here)
- security groups
- network ACLs
- internet gateways
- route tables
- network interfaces
- VPCs (above resources required to be deleted first)

As you can see, I based my work off of the structure you had already started in the original script. With the number of resources to be deleted, I'm duplicating a lot of code and am looking at ways to reduce that by generalizing the functions and passing in various parameters and possibly calling certain functions dynamically. That will be in a future version that removes the resources listed above and finally the VPCs that we create.

Otherwise, I think this script is in decent enough shape for a review and possibly deployment, if you agree. I have run it in a docker container from my laptop and cleaned up a number of unused resources already.

Let me know if you have any questions or feedback.

ShawonC commented on lambdas/ec2/reaper.py:
```diff
@@ -9,6 +9,9 @@
 import boto3
 
 ec2 = boto3.resource('ec2')
+ec2_client = boto3.client('ec2')
```
@greglarkin did you switch to using ec2_client here for consistency purposes? Or was there a technical reason as well?

shawon-crosen commented:
@greglarkin thanks for all the work on this, it's really great to see you managing other resources. 

I'd say this looks good enough to deploy for now, though I agree that for a final production version we should probably look to create something along the lines of a resource class with generic tag and terminator functions that get values passed from resource objects. 

In terms of this PR there are a couple things outstanding that need to be done before being able to merge:

1. The deploy_reaper.yaml cloudformation template needs updated to call the correct function in AWS Lambda: https://github.com/puppetlabs/aws_resource_reaper/blob/master/lambdas/ec2/deploy_reaper.yaml#L48
I don't know that just replacing 'reaper.terminate_expired_instances' with 'reaper.terminate_expired_resources' will work, as the 'reaper.terminate_expired_resources' function expects arguments. 

2. The slack_notifier.py script will need updated, though that should really just be adding empty alert message strings for each resource to the NO_ALERT list, and making sure events are getting sent correctly: https://github.com/puppetlabs/aws_resource_reaper/blob/master/lambdas/ec2/slack_notifier.py 

In the meantime, I can do a deploy of this into the puppetlabs-lms account and try to get the cloudformation template piece ironed out. I'll plan to get this update deployed by EOD Thursday.

Commit:
Add code to remove security groups and VPCs (eventually)

greglarkin commented:
Thanks for the feedback @ShawonC! To answer your questions:

1) AFAIK, the original entry point function took 2 parameters (CloudWatch schedule event parameters, maybe?) so I essentially created a new wrapper function with the same signature as the previous entry point. It didn't look like I needed to pass those parameters anywhere else in the script but I figured I needed to keep the signature the same, if that makes sense. So if that's true, maybe the line that you referenced above would just need the new function name?

2) Great, thanks for letting me know about that. I'll make changes to that script as well.

I'm still adding other resources to be cleaned up, and we also weren't completely tagging everything we could when standing up resources in AWS. That has been corrected in our Terraform code now, and I'll continue to add some more functions to clean up all of the VPC dependencies.

If I can get the script working with all of that extra code in it, then I'll maybe take some other time to refactor it into a more compact form.

greglarkin commented:
Thanks for catching that, Shawon. It looks like I could use a Boto3
collection instead of calling the describe_vpcs() function from the
ec2_client object. It seemed a little weird to have to create a separate
object. I'll modify the code to avoid the use of ec2_client where I can.

-Greg


*Greg Larkin*
Principal Field Solutions Developer
*Puppet, Inc.*

<https://learn.puppet.com/>


On Tue, Jun 11, 2019 at 4:22 PM Shawon Crosen <notifications@github.com>
wrote:

> *@ShawonC* commented on this pull request.
> ------------------------------
>
> In lambdas/ec2/reaper.py
> <https://github.com/puppetlabs/aws_resource_reaper/pull/18#discussion_r292665630>
> :
>
> > @@ -9,6 +9,9 @@
>  import boto3
>
>  ec2 = boto3.resource('ec2')
> +ec2_client = boto3.client('ec2')
>
> @greglarkin <https://github.com/greglarkin> did you switch to using
> ec2_client here for consistency purposes? Or was there a technical reason
> as well?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/puppetlabs/aws_resource_reaper/pull/18?email_source=notifications&email_token=ABEVVTV4BK4DBPZCN6N47LLP2AJTDA5CNFSM4HUO6G32YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3HPUJY#pullrequestreview-248445479>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABEVVTR4MBO4GPLCU4HNUBLP2AJTDANCNFSM4HUO6G3Q>
> .
>

Commit:
Added more resources to delete before deleting the VPC. Switched away
from using the ec2_client object in favor of collections on the ec2
object. Unable to test completely due to drift between the end_time
and terminate_date tags but will have to align them for a more complete
test.

Commit:
Fix the EC2 termination waiter so it eventually exits. No need to
wait on the network interface status.

Commit:
Update the reaper with the information Slack messages. Update the
Slack notifier to ignore messages when no load balancers have been
affected.

greglarkin commented:
@ShawonC I just pushed some changes to attempt cleaning up all resources before deleting the VPC. It still doesn't completely clean up the stack. There are some issues with route tables that I haven't figured out yet. 

There is a "main" route table that I cannot delete directly in the AWS console. It complains about something associated with it, but I haven't seen anywhere to disassociate it yet. A similar error is returned from boto3. But then if I delete the entire VPC it's contained in from the console, everything is cleaned up correctly.

So some additional sleuthing is needed here. But as far as I can, the script will remove a lot of other resources, primarily the ones that cost us money (instances and load balancers). I started updating the Slack notifier, but I wonder if all of that code should be added for everything resource type. I'm concerned about all of the duplication already, and info messages that are displayed at the end of each `terminate_expired_...` function will add a lot more. What do you think? Are those messages necessary or useful?

greglarkin commented:
I also removed the need for that `ec2_client` object. There is now one place that I use it locally to invoke the `get_waiter()` function when waiting for EC2 instances to terminate.

greglarkin commented:
This is the start of the code block that I'm wondering if it is useful to duplicate into every resource type termination function: https://github.com/puppetlabs/aws_resource_reaper/pull/18/files#diff-1bd8264017880e23ca6b706145ed0765R700

acidprime commented:
@ShawonC ping

shawon-crosen commented:
@acidprime thanks for the ping. I've created this ticket and will get this tested and reviewed this sprint: https://tickets.puppetlabs.com/browse/DATAOPS-1300

shawon-crosen commented:
@greglarkin @acidprime I started reviewing these changes this sprint and reading over Greg's notes about code duplication got me thinking about how to consolidate some of this. I decided to work on that, adapting the code from Greg into a class and main script format. 

This branch shows my work so far, which is really close to doing what Greg's changes to the script initially were doing: https://github.com/puppetlabs/aws_resource_reaper/tree/reaper_enhancements/lambdas/reaper

I've moved all the methods that interact with resources into a reaper_class.py file, and the enforce and termination portions are in reaper.py. This way we can instantiate a reaper object based on the service we want to work against, and have a generic set of logging messages that variables are passed to in reaper.py. I also consolidated all the delete functions for ec2 resource objects into a generic delete_ec2_resource() function to reduce some of the code duplication. It also allows us to potentially just add ec2 resources to the list of resources for termination. 

I've been testing this week and am really close to having these updates ready for prod. @greglarkin if you have the time I'd love for you to test this yourself locally. You can do so by setting the aws_access_key and aws_secret_key variables for the boto3 instantiation in reaper.py, then removing the event and context arguments from the resource_reaper() function and calling it at the bottom of the script. 

@greglarkin thank you so much for your work on this, I adapted pretty much all of the code you added in this branch for these changes, and it drastically reduced the time it would've taken if I'd started trying to do this myself. I'm hoping to have a PR ready for review on that branch by end of day today or tomorrow. Please let me know how local testing goes if you get a chance. 

You can see the messages getting sent to Slack in the #reaper-it-services channel. Right now there are some duplicate route_table termination messages due to an error in the code, but otherwise that's about how it would look running every hour.

greglarkin commented:
Hi @ShawonC, great work - thanks for updating that! I just tested the code by making the changes you suggested and running the following command as I had been a while back to invoke a lambda function locally:

`docker run --rm -v "$PWD":/var/task lambci/lambda:python3.7 reaper.resource_reaper`

I set `LIVEMODE` to `False` in the script to run in no-op mode and got the following output. Looks good to me while spot-checking it.

Did you find a specific reference describing the order that resources have to be torn down to delete a VPC? That was the most difficult thing to do, from what I remember. I like how you have been able to list everything in an array and iterate through it.

I probably shouldn't try running with `LIVEMODE` set to `True` until our weekly classes are completed to make sure I don't delete something accidentally. But I would like to run it live at some point to see how it works. Looks like there are a lot of volumes and snapshots to be cleaned up, and they don't have `termination_date` tags, so I wonder if we'll have to figure out a way to add that tag to them. Looks like they are costing us money to keep around.

[reaper.log](https://github.com/puppetlabs/aws_resource_reaper/files/3553180/reaper.log)

shawon-crosen commented:
@greglarkin good to know about that lambda docker container! I'll have to try that going forward for running locally. 

That log output is looking good. 

In regards to VPCs, all I found was that all resources attached to them need to be detached or deleted before the VPC can be deleted. The only things excluded from that are the default network acl and the main route table. I have the services set up so that the VPC gets deleted last compared to resources that would attach to it, so that should solve the problem. One thing I have noticed though is if a resource attached doesn't have a termination_date tag, like a subnet for example, the delete call for the VPC will fail because the subnet is still active. 

That's why I wrapped the delete_resource() function in a try/except block, so that if this does happen that error will at least print to the logs. 

Are you okay with me closing this PR and opening a new one against that new branch? 

Let me know if you would like to get it set up in AWS Lambda like the current reaper for LIVEMODE testing, or if I can help out in any other way with that.

shawon-crosen commented:
@greglarkin here's the link to the new PR against the new branch: https://github.com/puppetlabs/aws_resource_reaper/pull/19

I initially pulled in your changes so it will show your work as well in that PR.

greglarkin commented:
Superseded by #19
